contrib/dig/dig.1
.\"
.\" Distributed with 'dig' version 2.0 from University of Southern
.\" California Information Sciences Institute (USC-ISI).
.\"
.\"       dig.1   2.0 (USC-ISI) 8/30/90
.\"
.\" Man page reformatted for this release by Andrew Cherenson
.\" (arc@sgi.com)
.\"
.TH DIG 1 "August 30, 1990"
.SH NAME
dig \- send domain name query packets to name servers
.SH SYNOPSIS
.B dig 
.RI [ @\fIserver\fP ]
.I domain
.RI [ "<query-type>" ]
.RI [ "<query-class>" ]
.RI [ "+<query-option>" ]
.RI [ "\-<dig-option>" ]
.RI [ "%comment" ]
.SH DESCRIPTION
\fIDig\fP (domain information groper) is a flexible command line tool
which can be used to gather information from the Domain
Name System servers. \fIDig\fP has two modes: simple interactive mode
which makes a single query, and batch which executes a query for
each in a list of several query lines. All query options are
accessible from the command line.
.PP
The usual simple use of \fIdig\fP will take the form:
.sp 1
	dig  @server  domain   query-type  query-class
.sp 1
where:
.IP \fIserver\fP 
may be either a domain name or a dot-notation
Internet address. If this optional field is omitted, \fIdig\fP
will attempt to use the default name server for your machine.
.sp 1
\fBNote:\fP If a domain name is specified, this will be resolved
using the domain name system resolver (i.e., BIND). If your
system does not support DNS,  you may \fIhave\fP to specify a
dot-notation address.  Alternatively, if there is a server
at your disposal somewhere,  all that is required is that
/etc/resolv.conf be present and indicate where the default
name servers  reside,  so that  \fIserver\fP itself can be
resolved. See 
.IR resolver (5)
for information on /etc/resolv.conf.
(WARNING: Changing /etc/resolv.conf will affect
the standard  resolver library and  potentially several
programs which use it.) As an option, the user may set the
environment variable LOCALRES to name a file which is to
be used instead of /etc/resolv.conf (LOCALRES is specific
to the \fIdig\fP resolver and  not referenced by the standard
resolver). If the LOCALRES variable is not set or the file
is not readable then /etc/resolv.conf will be used.
.IP \fIdomain\fP
is the domain name for which you are requesting information.
See OPTIONS [-x] for convenient way to specify inverse address
query.
.IP \fIquery-type\fP 
is the type of information (DNS query type) that
you are requesting. If omitted, the default is "a" (T_A = address).
The following types are recognized:
.sp 1
.ta \w'hinfoXX'u +\w'T_HINFOXX'u
.nf
a	T_A	network address
any	T_ANY	all/any information about specified domain
mx	T_MX	mail exchanger for the domain
ns	T_NS	name servers
soa	T_SOA	zone of authority record
hinfo	T_HINFO	host information
axfr	T_AXFR	zone transfer
		 (must ask an authoritative server)
txt	T_TXT	arbitrary number of strings
		 (not yet supported by BIND)
.fi
.sp 1
(See RFC 1035 for the complete list.)
.IP \fIquery-class\fP
is the network class requested in the query. If
omitted, the default is "in" (C_IN = Internet).
The following classes are recognized:
.sp 1
.ta \w'hinfoXX'u +\w'T_HINFOXX'u
.nf
in	C_IN	Internet class domain
any	C_ANY	all/any class information
.fi
.sp 1
(See RFC 1035 for the complete list.)
.sp 1
\fBNote:\fP
"Any" can be used to specify a class and/or a type of
query. \fIDig\fP will parse the first occurrence of "any"
to mean query-type = T_ANY. To specify query-class =
C_ANY you must either specify "any" twice, or set
query-class using "\-c" option (see below).
.SH OTHER OPTIONS
.IP "%ignored-comment"
"%" is used to included an argument that is simply not
parsed.  This may be useful  if running \fIdig\fP in batch
mode. Instead of resolving every @server-domain-name in
a list of queries, you can avoid the overhead of doing
so, and still have the domain name on the command line
as a reference. Example:
.sp 1
	dig  @128.9.0.32  %venera.isi.edu  mx  isi.edu
.sp 1
.IP "\-<dig option>"
"\-" is used to specify an option which effects the
operation of \fIdig\fP. The following options are currently
available (although not guaranteed to be useful):
.RS
.IP "\-x \fIdot-notation-address\fP"
Convenient form to specify inverse address mapping.
Instead of "dig 32.0.9.128.in-addr.arpa" one can
simply "dig -x 128.9.0.32".
.IP "\-f \fIfile\fP"
File for \fIdig\fP batch mode. The file contains a list
of query specifications (\fIdig\fP command lines) which
are to be executed successively. Lines beginning
with ';', '#', or '\\n' are ignored. Other options
may still appear on command line, and will be in
effect for each batch query.
.IP "\-T \fItime\fP"
Time in seconds between start of successive
queries when running in batch mode. Can be used
to keep two or more batch \fIdig\fP commands running
roughly in sync. Default is zero.
.IP "\-p \fIport\fP"  
Port number. Query a name server listening to a
non-standard port number. Default is 53.
.IP "\-P[\fIping-string\fP]"
After query returns, execute a 
.IR ping (8)
command
for response time comparison. This rather
unelegantly makes a call to the shell. The last
three lines of statistics is printed for the
command:
.sp 1
	ping \-s server_name 56 3
.sp 1
If the optional "ping string" is present, it
replaces "ping \-s" in the shell command.
.IP "\-t \fIquery-type\fP"
Specify type of query. May specify either an
integer value to be included in the type field
or use the abbreviated mnemonic as discussed
above (i.e., mx  = T_MX).
.IP "\-c \fIquery-class\fP"  
Specify class of query. May specify either an
integer value to be included in the class field
or use the abbreviated mnemonic as discussed
above (i.e., in = C_IN).
.IP "\-envsav"
This flag specifies that the \fIdig\fP environment
(defaults, print options, etc.), after
all of the arguments are parsed, should be saved
to a file to become the default environment.
Useful if you do not like the standard set of
defaults and do not desire to include a
large number of options each time \fIdig\fP is used.
The environment consists of resolver state
variable flags, timeout, and retries as well as
the flags detailing \fIdig\fP output (see below).
If the shell environment variable LOCALDEF is set
to the name of a file, this is where the default
\fIdig\fP environment is saved. If not, the file
"DiG.env" is created in the current working directory.
.sp 1
\fBNote:\fP LOCALDEF is specific to the \fIdig\fP resolver,
and will not affect operation of the standard
resolver library.
.sp 1
Each time \fIdig\fP is executed, it looks for "./DiG.env"
or the file specified by the shell environment variable
LOCALDEF. If such file exists and is readable, then the
environment is restored from this file
before any arguments are parsed.
.IP "\-envset"
This flag only affects
batch query runs. When "\-envset" is
specified on a line in a \fIdig\fP batch file,
the \fIdig\fP environment after the arguments are parsed,
becomes the default environment for the duration of
the batch file, or until the next line which specifies
"\-envset".
.IP "\-[no]stick"
This flag only affects batch query runs.
It specifies that the \fIdig\fP environment (as read initially
or set by "\-envset" switch) is to be restored before each query
(line) in a \fIdig\fP batch file.
The default "\-nostick" means that the \fIdig\fP environment
does not stick, hence options specified on a single line
in a \fIdig\fP batch file will remain in effect for
subsequent lines (i.e. they are not restored to the
"sticky" default).

.RE
.IP "+<query option>"
"+" is used to specify an option to be changed in the
query packet or to change \fIdig\fP output specifics. Many
of these are the same parameters accepted by 
.IR nslookup (1).
If an option requires a parameter, the form is as
follows:
.sp 1
	+keyword[=value]
.sp 1
Most keywords can be abbreviated.  Parsing of the "+"
options  is very  simplistic \(em a value must not be
separated from its keyword by white space. The following
keywords are currently available:
.sp 1
.nf
.ta \w'domain=NAMEXX'u +\w'(deb)XXX'u
Keyword	Abbrev.	Meaning [default]

[no]debug	(deb)	turn on/off debugging mode [deb]
[no]d2		turn on/off extra debugging mode [nod2]
[no]recurse	(rec)	use/don't use recursive lookup [rec]
retry=#	(ret)	set number of retries to # [4]
time=#	(ti)	set timeout length to # seconds [4]
[no]ko		keep open option (implies vc) [noko]
[no]vc		use/don't use virtual circuit [novc]
[no]defname	(def)	use/don't use default domain name [def]
[no]search	(sea)	use/don't use domain search list [sea]
domain=NAME	(do)	set default domain name to NAME
[no]ignore	(i)	ignore/don't ignore trunc. errors [noi]
[no]primary	(pr)	 use/don't use primary server [nopr]
[no]aaonly	(aa)	authoritative query only flag [noaa]
[no]sort	(sor)	sort resource records [nosor]
[no]cmd		echo parsed arguments [cmd]
[no]stats	(st)	print query statistics (RTT,etc) [st]
[no]Header	(H)	print basic header [H]
[no]header	(he)	print header flags [he]
[no]ttlid	(tt)	print TTLs [tt]
[no]cl			print class info [nocl]
[no]qr		print outgoing query [noqr]
[no]reply	(rep)	print reply [rep]
[no]ques	(qu)	print question section [qu]
[no]answer	(an)	print answer section [an]
[no]author	(au)	print authoritative section [au]
[no]addit	(ad)	print additional section [ad]
pfdef		set to default print flags
pfmin		set to minimal default print flags
pfset=#		set print flags to #
		(# can be hex/octal/decimal)
pfand=#		bitwise and print flags with #
pfor=#		bitwise or print flags with #
.fi
.sp 1
The retry and time options affect the retransmission strategy used by resolver 
library when sending datagram queries. The algorithm is as follows:
.sp 1
.in +5n
.nf
for i = 0 to retry \- 1
    for j = 1 to num_servers
	send_query
	wait((time * (2**i)) / num_servers)
    end
end
.fi
.in -5n
.sp 1
(Note: \fIdig\fP always uses a value of 1 for num_servers.) 
.sp 1
Pfset, pfand, and pfor were included to make manipulation of
the various print options less tedious. Below are the currently
defined meanings for the various print flag bits.  They may be
combined (ANDed) to achieve various output formats.
.sp 1
.nf
.ta \w'PRF_HEADXXX'u +\w'0xXXXXXX'u
PRF_STATS	0x0001	RTT, query & server host, date, msg size
PRF_CLASS	0x0004	Resource record class information
PRF_CMD	0x0008	dig command line echo
PRF_QUES	0x0010	questions section
PRF_ANS	0x0020	answers section
PRF_AUTH	0x0040	authoritative section
PRF_ADD	0x0080	additional records section
PRF_HEAD1	0x0100	RR section headers & counts
PRF_HEAD2	0x0200	pkt header flags
PRF_TTLID	0x0400	Resource record ttl
PRF_HEADX	0x0800	basic header
PRF_QUERY	0x1000	outgoing query packet
PRF_REPLY	0x2000	reply packet
PRF_SORT	0x8000	sort various response sections
PRF_DEF	0x2ff9	default dig settings
PRF_ZONE	0x24f9	default setting for zone transfer
PRF_MIN	0xa930	minimalistic dig settings for
		(future) automated server testing
.fi
.sp 1
When setting the print options, if you want to see information other
than statistics, you should choose to examine the outgoing (0x1000), 
incoming (0x2000), or both packets plus the specific sections of the packet
you are interested in.
.SH DETAILS
\fIDig\fP requires a slightly modified version of the BIND 
.IR resolver (3)
library to gather count and time statistics. Otherwise, it is
straight-forward (albeit not pretty) effort of parsing arguments
and setting appropriate parameters. \fIDig\fP uses resolver routines
res_init(), res_mkquery(), res_send() as well as accessing _res
structure. Compiling \fIdig\fP with the standard resolver library
is possible, but will change the output format, make the print
options meaningless, and not gather RTT and packet count stats.
.SH FILES
.ta \w'/etc/resolv.confXX'u
/etc/resolv.conf	initial domain name and name server 
\./DiG.env		default save file for default options
.br
	addresses
.SH ENVIRONMENT
LOCALRES	file to use in place of /etc/resolv.conf
.br
LOCALDEF	default environment file
.SH AUTHOR
Steve Hotz 
hotz@isi.edu
.SH ACKNOWLEDGMENTS
\fIDig\fP uses functions from 
.IR nslookup (1)
authored by Andrew Cherenson;
taken from Berkeley BIND 4.8 distribution. The resolver library
is primarily from the Berkeley 4.8 BIND distribution.
.SH BUGS
\fIDig\fP has a serious case of "creeping featurism" -- the result of
considering several potential uses during it's development.  It would
probably benefit from a rigorous diet.  Similarly, the print flags
and granularity of the items they specify make evident their
rather ad hoc genesis.
.PP
\fIDig\fP does not consistently exit nicely (with appropriate status)
when a problem occurs somewhere in the resolver (NOTE: most of the common
exit cases are handled).  This is particularly annoying when running in
batch mode.  If it exits abnormally (and is not caught), the entire
batch aborts; when such an event is trapped, \fIdig\fP simply
continues with the next query.
.SH SEE ALSO
named(8),  resolver(3),  resolver(5),  nslookup(1)


contrib/dig/dig.c
/*	BSDI $Id: dig.c,v 1.2 1993/12/17 03:39:11 sanders Exp $	*/

 /*******************************************************************
 **      DiG -- Domain Information Groper                          **
 **                                                                **
 **        dig.c - Version 2.0 (9/1/90)                            **
 **                                                                **
 **        Developed by: Steve Hotz & Paul Mockapetris             **
 **        USC Information Sciences Institute (USC-ISI)            **
 **        Marina del Rey, California                              **
 **        1989                                                    **
 **                                                                **
 **        dig.c -                                                 **
 **           Version 2.0 (9/1/90)                                 **
 **               o renamed difftime() difftv() to avoid           **
 **                 clash with ANSI C                              **
 **               o fixed incorrect # args to strcmp,gettimeofday  **
 **               o incorrect length specified to strncmp          **
 **               o fixed broken -sticky -envsa -envset functions  **
 **               o print options/flags redefined & modified       **
 **                                                                **
 **           Version 2.0.beta (5/9/90)                            **
 **               o output format - helpful to `doc`               **
 **               o minor cleanup                                  **
 **               o release to beta testers                        **
 **                                                                **
 **           Version 1.1.beta (10/26/89)                          **
 **               o hanging zone transer (when REFUSED) fixed      **
 **               o trailing dot added to domain names in RDATA    **
 **               o ISI internal                                   **
 **                                                                **
 **           Version 1.0.tmp  (8/27/89)                           **
 **               o Error in prnttime() fixed                      **
 **               o no longer dumps core on large pkts             **
 **               o zone transfer (axfr) added                     **
 **               o -x added for inverse queries                   **
 **                               (i.e. "dig -x 128.9.0.32")       **
 **               o give address of default server                 **
 **               o accept broadcast to server @255.255.255.255    **
 **                                                                **
 **           Version 1.0  (3/27/89)                               **
 **               o original release                               **
 **                                                                **
 **     DiG is Public Domain, and may be used for any purpose as   **
 **     long as this notice is not removed.                        **
 ****                                                            ****
 ****   NOTE: Version 2.0.beta is not for public distribution    ****
 ****                                                            ****
 *******************************************************************/


#define VERSION 20
#define VSTRING "2.0"

#include "hfiles.h"

#include <sys/types.h>
#include <netinet/in.h>
#include <stdio.h>
#include RESOLVH
#include <sys/file.h>
#include <sys/stat.h>
#include <ctype.h> 
#include <errno.h>
#include "subr.c"
#include "pflag.h"
#include "options.c"

#ifdef RESLOCAL
#include "qtime.c"
#else
#include "qtime.h"
#endif

int pfcode = PRF_DEF;
int eecode = 0;
extern char *inet_ntoa();
extern struct state _res;

FILE  *qfp;
int sockFD;

#define MAXDATA		256   
#define SAVEENV "DiG.env"

char *defsrv, *srvmsg;
char defbuf[40] = "default -- ";
char srvbuf[60];

#define UC(b)   (((int)b)&0xff)

 /*
 ** Take arguments appearing in simple string (from file)
 ** place in char**.
 */
stackarg(y,l)
     char *l;
     char **y;
{
  int done=0;
  while (!done) {
    switch (*l) {
    case '\t':
    case ' ':  l++;    break;
    case NULL:
    case '\n': done++; *y = NULL; break;
    default:   *y++=l;  while (!isspace(*l)) l++;
      if (*l == '\n') done++; *l++ = NULL; *y = NULL;
}}}


main(argc, argv)
     int argc;
     char **argv;
{
	struct hostent *hp;
	short port = htons(NAMESERVER_PORT);
	char packet[PACKETSZ];
	char answer[PACKETSZ];
	int n;
	char doping[90];
	char pingstr[50];
	char *afile;
	unsigned long tmpaddr;
        char revaddr[10][10];
        int addri, addrj;
	char *addrc;

	struct timeval exectime, tv1,tv2,tv3;
	char curhost[30];

	char *srv;
	int anyflag = 0;
	int sticky = 0;
	int tmp; 
	int qtype = 1, qclass = 1;
	int addrflag = 0;
	int zone = 0;
        int bytes_out, bytes_in;

	char cmd[256];
	char domain[128];
        char msg[120], *msgptr;
	char **vtmp;
	char *args[30];
	char **ax;
	char **ay;
	int once = 1, dofile=0; /* batch -vs- interactive control */
	char fileq[100];
	char *qptr;
	int  fp;
	int wait=0;
	int envset=0, envsave=0;
	int Xpfcode, Tpfcode, Toptions;
	char *Xres, *Tres;
	char *pp;


	res_init();
	gethostname(curhost,30);
	defsrv = strcat(defbuf, inet_ntoa(_res.nsaddr.sin_addr));
	_res.options |= RES_DEBUG;
	Xres = (char *) malloc(sizeof(_res)+1);
	Tres = (char *) malloc(sizeof(_res)+1);
	bcopy(&_res,Xres,sizeof(struct state));
	Xpfcode=pfcode;

 /*
 ** If LOCALDEF in environment, should point to file
 ** containing local favourite defaults.  Also look for file
 ** DiG.env (i.e. SAVEENV) in local directory.
 */

	if ((((afile= (char *) getenv("LOCALDEF")) != (char *) NULL) &&
	     ((fp=open(afile,O_RDONLY)) > 0)) ||
	    ((fp = open(SAVEENV,O_RDONLY)) > 0)) {
	  read(fp,Xres,sizeof(struct state));
	  read(fp,&Xpfcode,sizeof(int));
	  close(fp);
	  bcopy(Xres,&_res,sizeof(struct state));
	  pfcode = Xpfcode;
	}
 /*
 **   check for batch-mode DiG 
 */
	vtmp = argv; ax=args;
	while (*vtmp != NULL) {
	  if (strcmp(*vtmp,"-f") == 0) {
	    dofile++; once=0;
	    if ((qfp = fopen(*++vtmp,"r")) == NULL) {
	      fflush(stdout);
	      perror("file open");
	      fflush(stderr);
	      exit(10);
	    }
	  } else {
	    *ax++ = *vtmp;
	  }
	  vtmp++;
	}

	_res.id = 1;
	gettimeofday(&tv1,NULL);

 /*
 **  Main section: once if cmd-line query
 **                while !EOF if batch mode
 */
	*fileq=NULL;
	while ((dofile && (fgets(fileq,100,qfp) !=NULL)) || 
	       ((!dofile) && (once--))) 
	  {
	    if ((*fileq=='\n') || (*fileq=='#') || (*fileq==';'))
	      continue; /* ignore blank lines & comments */

/*
 * "sticky" requests that before current parsing args
 * return to current "working" environment (X******)
 */
	    if (sticky) {
	      bcopy(Xres,&_res,sizeof(struct state));
	      pfcode = Xpfcode;
	    }

/* concat cmd-line and file args */
	    ay=ax;
	    qptr=fileq;
	    stackarg(ay,qptr);

	    /* defaults */
	    qtype=qclass=1;
	    zone = 0;
	    *pingstr=0;
	    srv=NULL;

	    sprintf(cmd,"\n; <<>> DiG %s <<>> ",VSTRING);
	    argv = args;
/*
 * More cmd-line options than anyone should ever have to
 * deal with ....
 */
	    while (*(++argv) != NULL) { 
	      strcat(cmd,*argv); strcat(cmd," ");
	      if (**argv == '@') {
		srv = (*argv+1);
		continue;
	      }
	      if (**argv == '%')
		continue;
	      if (**argv == '+') {
		SetOption(*argv+1);
		continue;
	      }
	 
	      if (strncmp(*argv,"-nost",5) == 0) {
		sticky=0;; continue;
	      } else if (strncmp(*argv,"-st",3) == 0) {
		sticky++; continue;
	      } else if (strncmp(*argv,"-envsa",6) == 0) {
		envsave++; continue;
	      } else if (strncmp(*argv,"-envse",6) == 0) {
		envset++; continue;
	      }

         if (**argv == '-') {
	   switch (argv[0][1]) { 
	   case 'T': wait = atoi(*++argv);
	     break;
	   case 'c': 
	     if ((tmp = atoi(*++argv)) || *argv[0]=='0') {
	       qclass = tmp;
	     } else if (tmp = StringToClass(*argv,0)) {
	       qclass = tmp;
	     } else {
	       printf("; invalid class specified\n");
	     }
	     break;
	   case 't': 
	     if ((tmp = atoi(*++argv)) || *argv[0]=='0') {
	       qtype = tmp;
	     } else if (tmp = StringToClass(*argv,0)) {
	       qtype = tmp;
	     } else {
	       printf("; invalid type specified\n");
	     }
	     break;
	   case 'x':
	     if (qtype == T_A) qtype = T_ANY;
	     if ((addrc = *++argv) == (char *)0) {
	       printf(";; -x requires an argument\n");
	       continue;
	     }
	     addri=addrj=0;
	     while (*addrc) {
	       if (*addrc == '.') {
		 revaddr[addri][addrj++] = '.';
		 revaddr[addri][addrj] = (char) 0;
		 addri++; addrj=0;
	       } else {
		 revaddr[addri][addrj++] = *addrc;
	       }
	       addrc++;
	     }
	     if (*(addrc-1) == '.') {
	       addri--;
	     } else {
	       revaddr[addri][addrj++] = '.';
	       revaddr[addri][addrj] = (char) 0;
	     }
	     *domain = (char) 0;
	     for (addrj=addri; addrj>=0; addrj--)
	       strcat(domain,revaddr[addrj]);
	     strcat(domain,"in-addr.arpa.");
/* old code -- some bugs
	     tmpaddr = inet_addr(*++argv);
	     pp = (char *) &tmpaddr;
	     sprintf(domain,"%d.%d.%d.%d.in-addr.arpa",
		     UC(pp[3]), UC(pp[2]), UC(pp[1]), UC(pp[0]));
*/        
	     break;
	   case 'p': port = htons(atoi(*++argv)); break;
	   case 'P':
	     if (argv[0][2] != NULL)
	       strcpy(pingstr,&argv[0][2]);
	     else
	       strcpy(pingstr,"ping -s");
	     break;
	   } /* switch - */
	   continue;
	 } /* if '-'   */

	      if ((tmp = StringToType(*argv,-1)) != -1) { 
		if ((T_ANY == tmp) && anyflag++) {  
		  qclass = C_ANY; 	
		  continue; 
		}
		if (T_AXFR == tmp) {
		  pfcode = PRF_ZONE;
		  zone++;
		} else {
		  qtype = tmp; 
		}
	      }
	      else if ((tmp = StringToClass(*argv,-1)) != -1) { 
		qclass = tmp; 
	      }	 else {
		bzero(domain,128);
		sprintf(domain,"%s",*argv);
	      }
	    } /* while argv remains */
if (pfcode & 0x80000)
  printf("; pflag: %x res: %x\n", pfcode, _res.options);
	  
/*
 * Current env. (after this parse) is to become the
 * new "working environmnet. Used in conj. with sticky.
 */
	    if (envset) {
	      bcopy(&_res,Xres,sizeof(struct state));
	      Xpfcode=pfcode;
	      envset=0;
	    }

/*
 * Current env. (after this parse) is to become the
 * new default saved environmnet. Save in user specified
 * file if exists else is SAVEENV (== "DiG.env").
 */
	    if (envsave) {
	      if ((((afile= (char *) getenv("LOCALDEF")) != (char *) NULL) &&
		   ((fp=open(afile,O_WRONLY|O_CREAT|O_TRUNC,
			     S_IREAD|S_IWRITE)) > 0)) ||
		  ((fp = open(SAVEENV,O_WRONLY|O_CREAT|O_TRUNC,
			       S_IREAD|S_IWRITE)) > 0)) {
		write(fp,&_res,sizeof(struct state));
		write(fp,&pfcode,sizeof(int));
		close(fp);
	      }
	      envsave=0;
	    }

	    if (pfcode & PRF_CMD)
	      printf("%s\n",cmd);

      addrflag=anyflag=0;

/*
 * Find address of server to query. If not dot-notation, then
 * try to resolve domain-name (if so, save and turn off print 
 * options, this domain-query is not the one we want. Restore
 * user options when done.
 * Things get a bit wierd since we need to use resolver to be
 * able to "put the resolver to work".
 */

   srvbuf[0]=0;
   srvmsg=defsrv;
   if (srv != NULL) {
     tmpaddr = inet_addr(srv);
     if ((tmpaddr != (unsigned)-1) || 
	 (strncmp("255.255.255.255",srv,15) == 0)) {
       _res.nscount = 1;
       _res.nsaddr.sin_addr.s_addr = tmpaddr;
       srvmsg = strcat(srvbuf, srv);
     } else {
      Tpfcode=pfcode;
      pfcode=0;
      bcopy(&_res,Tres,sizeof(_res));
      Toptions = _res.options;
      _res.options = RES_DEFAULT;
      res_init();
      hp = gethostbyname(srv);
      pfcode=Tpfcode;
      if (hp == NULL) {
	fflush(stdout);
	fprintf(stderr,
		"; Bad server: %s -- using default server and timer opts\n",
		srv);
	fflush(stderr);
        srvmsg = defsrv;
	srv = NULL;
	_res.options = Toptions;
      } else {
	bcopy(Tres,&_res,sizeof(_res));
	bcopy(hp->h_addr, &_res.nsaddr_list[0].sin_addr, hp->h_length);
	_res.nscount = 1;
        srvmsg = strcat(srvbuf,srv);
        strcat(srvbuf, "  ");
        strcat(srvmsg,inet_ntoa(_res.nsaddr.sin_addr));
      }
    }
     _res.id += _res.retry;
   }

       _res.id += _res.retry;
/*       _res.options |= RES_DEBUG;  */
       _res.nsaddr.sin_family = AF_INET;
       _res.nsaddr.sin_port = port;

       if (zone) {
	 do_zone(domain,qtype);
	 if (pfcode & PRF_STATS) {
	   gettimeofday(&exectime,NULL);
	   printf(";; FROM: %s to SERVER: %s\n",curhost,srvmsg);
	   printf(";; WHEN: %s",ctime(&(exectime.tv_sec)));
	 }
	  
	 fflush(stdout);
	 continue;
       }

       bytes_out = n = res_mkquery(QUERY, domain, qclass, qtype,
                         (char *)0, 0, NULL, packet, sizeof(packet));
       if (n < 0) {
	 fflush(stderr);
	 printf(";; res_mkquery: buffer too small\n\n");
	 continue;
       }
       eecode = 0;
       if ((bytes_in = n = res_send(packet, n, answer, sizeof(answer))) < 0) {
	 fflush(stdout);
         n = 0 - n;
         msg[0]=0;
         strcat(msg,";; res_send to server ");
         strcat(msg,srvmsg);
	 perror(msg);
	 fflush(stderr);

#ifndef RESLOCAL
/*
 * resolver does not currently calculate elapsed time
 * if there is an error in res_send()
 */
/*
	 if (pfcode & PRF_STATS) {
	   printf(";; Error time: "); prnttime(&hqtime); putchar('\n');
	 }
*/
#endif
	 if (!dofile) {
            if (eecode)
	      exit(eecode);
	    else
	      exit(9);
	  }
       }
#ifndef RESLOCAL
       else {
	 if (pfcode & PRF_STATS) {
	   printf(";; Sent %d pkts, answer found in time: ",hqcount);
	   prnttime(&hqtime);
	   if (hxcount)
	     printf(" sent %d too many pkts",hxcount);
	   putchar('\n');
	 }
       }
#endif

       if (pfcode & PRF_STATS) {
	 gettimeofday(&exectime,NULL);
	 gethostname(curhost,30);
	 printf(";; FROM: %s to SERVER: %s\n",curhost,srvmsg);
	 printf(";; WHEN: %s",ctime(&(exectime.tv_sec)));
         printf(";; MSG SIZE  sent: %d  rcvd: %d\n", bytes_out, bytes_in);
       }
	  
       fflush(stdout);
/*
 *   Argh ... not particularly elegant. Should put in *real* ping code.
 *   Would necessitate root priviledges for icmp port though!
 */
       if (*pingstr) {
         sprintf(doping,"%s %s 56 3 | tail -3",pingstr,
		 (srv==NULL)?(defsrv+10):srv);
	 system(doping);
       }
       putchar('\n');

/*
 * Fairly crude method and low overhead method of keeping two
 * batches started at different sites somewhat synchronized.
 */
       gettimeofday(&tv2, NULL);
       tv1.tv_sec += wait;
       difftv(&tv1,&tv2,&tv3);
       if (tv3.tv_sec > 0)
	 sleep(tv3.tv_sec);
	  }
	return(eecode);
      }
